### PR TITLE
Improve the reaction of IPv6 SLAAC to renumbering events

### PR DESCRIPTION
**Pull request by fgont**
PIO lifetimes are capped as follows:
 * Preferred Lifetime: capped to Router Lifetime
 * Valid Lifetime: capped to 48 * Router Lifetime

RIO lifetimes are capped to the Router Lifetime

Please see: https://tools.ietf.org/html/draft-gont-6man-slaac-renum-05
Problem statement: https://tools.ietf.org/html/draft-ietf-v6ops-slaac-renum-01

### CODE REVIEW

File: src/network/networkd-ndisc.c [hunk 1]
```diff
@@ -508,7 +508,7 @@ static int ndisc_router_process_route(Link *link, sd_ndisc_router *rt) {
 
         r = sd_ndisc_router_route_get_lifetime(rt, &lifetime);
         if (r < 0)
-                return log_link_warning_errno(link, r, "Failed to get gateway address from RA: %m");
+                return log_link_warning_errno(link, r, "Failed to get route lifetime from RA: %m");
```
Comment (ssahani) [inline src/network/networkd-ndisc.c:512]: Please drop this.

File: src/network/networkd-ndisc.c [hunk 2]
```diff
@@ -344,6 +346,7 @@ static int ndisc_router_generate_addresses(Link *link, unsigned prefixlen, uint3
 }
 
 static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *rt) {
+        uint16_t router_lifetime;
         uint32_t lifetime_valid, lifetime_preferred, lifetime_remaining;
```
Comment (ssahani) [inline src/network/networkd-ndisc.c:349]: reverse tree order

**Comment by fgont** [comment]
On 29/3/20 23:26, Susant Sahani wrote:
> *@ssahani* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/network/networkd-ndisc.c 
> <https://github.com/systemd/systemd/pull/15260#discussion_r399896547>:
> 
>> @@ -508,7 +508,7 @@ static int ndisc_router_process_route(Link *link, sd_ndisc_router *rt) {
>   
>           r = sd_ndisc_router_route_get_lifetime(rt, &lifetime);
>           if (r < 0)
> -                return log_link_warning_errno(link, r, "Failed to get gateway address from RA: %m");
> +                return log_link_warning_errno(link, r, "Failed to get route lifetime from RA: %m");
> 
> Please drop this.

Why?




-- 
Fernando Gont
SI6 Networks
e-mail: fgont@si6networks.com
PGP Fingerprint: 6666 31C6 D484 63B2 8FB1 E3C4 AE25 0D55 1D4E 7492

**Comment by fgont** [comment]
On 29/3/20 23:26, Susant Sahani wrote:
> *@ssahani* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/network/networkd-ndisc.c 
> <https://github.com/systemd/systemd/pull/15260#discussion_r399896650>:
> 
>> @@ -344,6 +346,7 @@ static int ndisc_router_generate_addresses(Link *link, unsigned prefixlen, uint3
>   }
>   
>   static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *rt) {
> +        uint16_t router_lifetime;
>           uint32_t lifetime_valid, lifetime_preferred, lifetime_remaining;
> 
> reverse tree order

Could you please elaborate?


-- 
Fernando Gont
SI6 Networks
e-mail: fgont@si6networks.com
PGP Fingerprint: 6666 31C6 D484 63B2 8FB1 E3C4 AE25 0D55 1D4E 7492

File: src/network/networkd-ndisc.c [hunk 3]
```diff
@@ -368,10 +371,19 @@ static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *r
         if (r < 0)
                 return log_link_error_errno(link, r, "Failed to get prefix valid lifetime: %m");
 
+        r = sd_ndisc_router_get_lifetime(rt, &router_lifetime);
+        if (r < 0)
```
Comment (ssahani) [inline src/network/networkd-ndisc.c]: Does this value always appears or can be ignored ? If it's not mandatory then we should not return here at all.  Cap should be calculated only if this value is present or the code the the same as it is.

File: src/network/networkd-ndisc.c [hunk 4]
```diff
@@ -444,6 +457,16 @@ static int ndisc_router_process_onlink_prefix(Link *link, sd_ndisc_router *rt) {
         if (r < 0)
                 return log_link_error_errno(link, r, "Failed to get prefix lifetime: %m");
 
+        r = sd_ndisc_router_get_lifetime(rt, &router_lifetime);
+        if (r < 0)
```
Comment (ssahani) [inline src/network/networkd-ndisc.c]: same as above . We should not return if this value is not here and do the same as it was doing before.

File: src/network/networkd-ndisc.c [hunk 5]
```diff
@@ -425,6 +437,7 @@ static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *r
 static int ndisc_router_process_onlink_prefix(Link *link, sd_ndisc_router *rt) {
         _cleanup_(route_freep) Route *route = NULL;
         usec_t time_now;
+        uint16_t router_lifetime;
```
Comment (ssahani) [inline src/network/networkd-ndisc.c:443]: Reverse tree order

**Comment by ssahani** [comment]
CI Failed needs rework.

File: src/network/networkd-ndisc.c (hunk 1 above)
Comment (fgont) [reply src/network/networkd-ndisc.c:512]: Done as you suggested.

File: src/network/networkd-ndisc.c (hunk 2 above)
Comment (fgont) [reply src/network/networkd-ndisc.c:349]: Fixed

File: src/network/networkd-ndisc.c (hunk 3 above)
Comment (fgont) [reply src/network/networkd-ndisc.c]: DOnes as suggested.

File: src/network/networkd-ndisc.c (hunk 5 above)
Comment (fgont) [reply src/network/networkd-ndisc.c:443]: Same

File: src/network/networkd-ndisc.c [hunk 6]
```diff
@@ -372,6 +375,19 @@ static int ndisc_router_process_autonomous_prefix(Link *link, sd_ndisc_router *r
...
                 return log_link_error_errno(link, r, "Failed to get prefix preferred lifetime: %m");
 
+        /* Lifetime of Autoconf prefixes are capped to improve the reaction of SLAAC to renumbering events.
+           Valid Lifetime: capped to Router Lifetime * DFLT_VLTIME_MULT
+           Preferred Lifetime: capped to Router Lifetime
+           See draft-gont-6man-slaac-renum
+         */
+        r = sd_ndisc_router_get_lifetime(rt, &router_lifetime);
+        if (r >= 0){
+                if(router_lifetime * DFLT_VLTIME_MULT < NDISC_PREFIX_LFT_MAX)
```
Comment (ssahani) [inline src/network/networkd-ndisc.c:385]: space `if (r >= 0) {`

**Comment by ssahani** [comment]
Please squash the review comments with the original PR.

**Comment by fgont** [comment]
On 30/3/20 03:04, Susant Sahani wrote:
> Please squash the review comments with the original PR.

Space fixed, and commits squashed. Ok?

Thanks,
-- 
Fernando Gont
SI6 Networks
e-mail: fgont@si6networks.com
PGP Fingerprint: 6666 31C6 D484 63B2 8FB1 E3C4 AE25 0D55 1D4E 7492

**Comment by ssahani** [comment]
> On 30/3/20 03:04, Susant Sahani wrote: Please squash the review comments with the original PR.
> Space fixed, and commits squashed. Ok? Thanks, -- Fernando Gont SI6 Networks e-mail: fgont@si6networks.com PGP Fingerprint: 6666 31C6 D484 63B2 8FB1 E3C4 AE25 0D55 1D4E 7492

Thank you . Let's wait for CI.

**Comment by fgont** [comment]
Anything more needed from my side?

**Comment by toreanderson** [comment]
Merging this is premature. The *slaac-renum* draft is at this point in time merely a *proposal* on how to update the IPv6 protocol standards; do ***not*** mistake an internet-draft for an accepted protocol update. (Anyone can post an internet-draft.)

Furthermore, the recommendations in section 4.1.2 of the draft regarding address lifetime capping are in fact incompatible with _current_ IPv6 standards in at least two ways. They break infinite lifetimes ([RFC4861 section 4.6.2](https://tools.ietf.org/html/rfc4861#section-4.6.2)), and the usage of SLAAC for address assignment when the advertising router is not a default router and therefore advertises AdvDefaultLifetime=0 (consider the use case where the RA includes [Route Information Options](https://tools.ietf.org/html/rfc4191#section-2.3) in order to advertise reachability to a limited number of destinations instead of a default route).

Therefore I would recommend against implementing these changes in systemd, until the proposed changes have been officially published in a standards track RFC.

File: src/network/networkd-ndisc.c [hunk 7]
```diff
@@ -19,6 +19,8 @@
 #define NDISC_DNSSL_MAX 64U
 #define NDISC_RDNSS_MAX 64U
 #define NDISC_PREFIX_LFT_MIN 7200U
+#define NDISC_PREFIX_LFT_MAX ((uint64_t) 0xffffffff)
```
Comment (poettering) [inline src/network/networkd-ndisc.c:22]: UINT64_C(0xffffffff)

File: src/network/networkd-ndisc.c [hunk 8]
```diff
@@ -19,6 +19,8 @@
...
+#define DFLT_VLTIME_MULT ((uint64_t) 48)
```
Comment (poettering) [inline src/network/networkd-ndisc.c:23]: ditto

File: src/network/networkd-ndisc.c (hunk 6 above)
Comment (poettering) [reply src/network/networkd-ndisc.c:385]: not addressed yet

**Comment by yuwata** [comment]
Replaced by #19639. Closing, as the document referred by this change is outdated.

**Pull request closed without merging**